### PR TITLE
[SDPA] Change autocast behavior

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Currenlty we autocast the top level SDPA composite funciton. This means that every subsequent ops is done in the lower precision type for the math path. We likely want the softmax to be done in higher precision. This chanages makes that so:
```
$3: f32[1, 1, 3, 6] = aten.mul.Scalar($0, 0.6389431042462725)
$4: f32[1, 1, 6, 5] = aten.transpose.int($1, -2, -1)
$5: f32[1, 1, 6, 5] = aten.mul.Scalar($4, 0.6389431042462725)
$6: f16[1, 1, 3, 6] = aten._to_copy.default($3, dtype=torch.float16)
$7: f16[1, 1, 6, 5] = aten._to_copy.default($5, dtype=torch.float16)
$8: f16[1, 1, 3, 6] = aten.expand.default($6, [1, 1, 3, 6])
$9: f16[1, 3, 6] = aten.view.default($8, [1, 3, 6])
$10: f16[1, 1, 6, 5] = aten.expand.default($7, [1, 1, 6, 5])
$11: f16[1, 6, 5] = aten.view.default($10, [1, 6, 5])
$12: f16[1, 3, 5] = aten.bmm.default($9, $11)
$13: f16[1, 1, 3, 5] = aten._unsafe_view.default($12, [1, 1, 3, 5])
$14: f32[1, 1, 3, 5] = aten._softmax.default($13, -1, True)
$15: f16[1, 1, 3, 5] = aten._to_copy.default($14, dtype=torch.float16)
$16: f16[1, 1, 5, 6] = aten._to_copy.default($2, dtype=torch.float16)
$17: f16[1, 1, 3, 5] = aten.expand.default($15, [1, 1, 3, 5])
$18: f16[1, 3, 5] = aten.view.default($17, [1, 3, 5])
$19: f16[1, 1, 5, 6] = aten.expand.default($16, [1, 1, 5, 6])
$20: f16[1, 5, 6] = aten.view.default($19, [1, 5, 6])
$21: f16[1, 3, 6] = aten.bmm.default($18, $20)
```

However dispatchign and alignment will be checked on the higher precision type and not after casting so we need to do more work to check that

cc @mcarilli @ptrblck @leslie-fang-intel @jgong5

### CODE REVIEW

**Comment by cpuhrsch** [comment]
I don't think this is needed because Softmax seems to accumulate in the higher precision type

See 
1) https://github.com/pytorch/pytorch/blob/cea899cd5781787c279ab6f5a523ed7709e33df1/aten/src/ATen/AccumulateType.h#L66-L68
2) https://github.com/pytorch/pytorch/blob/cea899cd5781787c279ab6f5a523ed7709e33df1/aten/src/ATen/native/cuda/PersistentSoftmax.cuh#L148
3) https://github.com/pytorch/pytorch/blob/cea899cd5781787c279ab6f5a523ed7709e33df1/aten/src/ATen/native/cuda/SoftMax.cu#L730

**Comment by drisspg** [comment]
Close because this isn't needed
